java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.AccessPath
+import arcs.core.data.expression.Expression
+
+internal typealias Path = List<AccessPath.Selector.Field>
+internal typealias Paths = List<Path>
+internal typealias ClaimDerivations = Map<Path, Set<Path>>
+
+/** A visitor that accumulates field [Paths] from a Paxel [Expression]. */
+class ExpressionPathAccumulator : Expression.Visitor<Paths> {
+
+    override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>): Paths {
+        return expr.left.accept(this) + expr.right.accept(this)
+    }
+
+    override fun <T> visit(expr: Expression.FieldExpression<T>): Paths {
+        if (expr.qualifier == null) {
+            return listOf(listOf(AccessPath.Selector.Field(expr.field)))
+        }
+        return listOf(
+            expr.qualifier!!.accept(this).first() + listOf(AccessPath.Selector.Field(expr.field))
+        )
+    }
+
+    override fun visit(expr: Expression.NewExpression): Paths =
+        expr.fields.map { it.second.accept(this) }.reduce { acc, set -> acc + set }
```

Comment:
Is this what you are doing? (Please check if the logic is correct.)
```suggestion
        expr.fields.flatMap { (_, expression) -> expression.accept(this) }.toSet()
```

Reply:
Double checked it -- a flatMap will work, but I don't need the `toSet()`.